Request.destination value tests

yoavweiss opened the pull request:
Solves #8810.

This is not yet a complete test suite, but as I'm having trouble running this locally, I want to see these pass on the CI bots before creating more test cases.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

yoavweiss committed:
Added tests that make sure img,script and style have the right destination value

w3c-bots commented:
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/22201)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/322432618)

yoavweiss committed:
Tested the tests

yoavweiss committed:
Removed equivalent tests in the preload directory

yoavweiss committed:
Added video and audio tests

yoavweiss commented:
This is now ready to review

jstenback commented on fetch/api/request/destination/fetch-destination-style.https.html:
```diff
@@ -0,0 +1,32 @@
+              assert_unreached("Fetch destination preload errored");
+          }));
+          link.href = "resources/dummy?dest=style";
+          frame.contentWindow.document.body.appendChild(script);
```
I'm assuming you meant to append 'link' here, not 'script'?

yoavweiss replied:
Yeah, in the latest version of the tests, it's all just `node`

wanderview commented on fetch/api/request/destination/fetch-destination-script.https.html:
```diff
@@ -0,0 +1,31 @@
+          frame.contentWindow.document.body.appendChild(node);
+        }))
+      .catch(unreached_rejection(t));
+}, 'Fetch destination script');
```
The "script" destination can also be produced by worker `importScripts()`.  We should have a test for that as well.

wanderview commented on fetch/api/request/destination/fetch-destination-style.https.html:
```diff
@@ -0,0 +1,32 @@
+          frame.contentWindow.document.body.appendChild(node);
+        }))
+      .catch(unreached_rejection(t));
+}, 'Fetch destination style');
```
We should have a test for CSS `@import` which can produce a "style" destination as well.

wanderview commented on fetch/api/request/destination/fetch-destination-image.https.html:
```diff
@@ -0,0 +1,30 @@
+          node.src = "dummy.png?dest=image";
+        }))
+      .catch(unreached_rejection(t));
+}, 'Fetch destination image');
```
Also imageset and picture elements.

wanderview reviewed (changes requested):
This is a good start, but its missing a lot of coverage.  The big challenge with a feature like destination is trying to make sure you cover all the various network requests the browser can generate.  I commented on some that are missing below, but we also need tests for these destination types:

* report
* document
* font
* sharedworker
* worker
* track
* manifest
* xslt

Not sure if we need "embed" or "object" since in theory those systems are being deprecated, I think.

Basically, we should try to test as many of the items in the table in the spec here:

https://fetch.spec.whatwg.org/#concept-request-destination

yoavweiss commented on fetch/api/request/destination/fetch-destination-audio.https.html:
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="../../../../service-workers/service-worker/resources/test-helpers.sub.js"></script>
```
I feel like these tests are better here, as they test the Fetch API *using Service Workers*. So I prefer to keep them here. Let me know if you vehemently disagree :)

yoavweiss commented on fetch/api/request/destination/fetch-destination-fetch.https.html:
```diff
@@ -0,0 +1,36 @@
+          frame.contentWindow.registration = registration;
+          var node = frame.contentWindow.document.createElement("script");
+          var text = frame.contentWindow.document.createTextNode(
+              'fetch("dummy?dest=").then(function(response) { ' +
```
Trying that, but I'm getting a "TypeError: Failed to fetch"...

wanderview replied:
That doesn't sound right.  We do that all the time in other service worker tests:

https://searchfox.org/mozilla-central/source/testing/web-platform/tests/service-workers/service-worker/fetch-event-after-navigation-within-page.https.html#23

yoavweiss replied:
Got it working

yoavweiss committed:
Refactored to promise based tests and addressed some review comments

yoavweiss commented on fetch/api/request/destination/resources/fetch-destination-worker.js:
```diff
@@ -0,0 +1,10 @@
+        if (event.request.destination == destination)
+            event.respondWith(fetch(event.request));
+        else
+            event.respondWith(Response.error());
```
sure

yoavweiss commented on fetch/api/request/destination/resources/fetch-destination-worker.js:
```diff
@@ -0,0 +1,10 @@
+self.addEventListener('fetch', function(event) {
+    if (event.request.url.indexOf('dummy') != -1) {
```
indeed

yoavweiss commented on fetch/api/request/destination/fetch-destination-audio.https.html:
```diff
@@ -0,0 +1,32 @@
+          var node = frame.contentWindow.document.createElement("audio");
+          node.addEventListener("loadstart", t.step_func_done(function() { registration.unregister(); }));
+          node.addEventListener("error", t.step_func_done(function() {
+              registration.unregister();
```
ok

yoavweiss commented on fetch/api/request/destination/fetch-destination-audio.https.html:
```diff
@@ -0,0 +1,32 @@
+<script src="../../../../service-workers/service-worker/resources/test-helpers.sub.js"></script>
+<script src="/common/media.js"></script>
+<script>
+async_test(function(t) {
```
refactored

yoavweiss commented:
@wanderview I addressed your comments on the existing tests. If possible I prefer to land this subset soon, and continue working on adding the missing destinations (that you rightfully pointed out) in separate patches.

wanderview commented on fetch/api/request/destination/fetch-destination.https.html:
```diff
@@ -0,0 +1,134 @@
+}, 'fetch() fetches with an empty string Request.destination');
+
+// HTMLAudioElement - audio destination
+promise_test(async t => {
```
nit: Thanks for trying to use more async functions as I requested.  For the cases where you have a single `await new Promise()` in the function you could probably just switch to `promise_test(new Promise()` instead.  What you have here works, though, and more just mentioning this in case you felt forced to use the await per my previous comment.

wanderview reviewed (approved):
Thanks!  LGTM, but as we discussed offline it would be great to have more test coverage before shipping the feature.

yoavweiss commented:
Thanks for reviewing! :)
